torch/optim/adam.py

```diff
@@ -37,14 +37,15 @@ def __init__(self, params, lr=1e-3, betas=(0.9, 0.999), eps=1e-8,
                 raise RuntimeError("`fused` does not support `differentiable`")
             self._step_supports_amp_scaling = True
             # TODO(crcrpar): [low prec params & their higher prec copy]
-            # Suppor AMP with FP16/BF16 model params which would need
+            # Support AMP with FP16/BF16 model params which would need
             # higher prec copy of params to do update math in higher prec to
             # alleviate the loss of information.
             if not all(
-                p.is_cuda and torch.is_floating_point(p)
```

Comment:
So this is in particular is not just for foreach but the fused adam kernel, which requires cuda (torch.fused_adam is only defined for CUDA). The check here would not be appropriate.

Reply:
> particular

@janeyx99  yeah, we are cuda-like device, so we will support the foreach-op and fused adam/adamw.

Reply:
Right, but the fused_adam op has stricter requirements than the foreach and is fundamentally different. I would prefer not conflating the two. Instead, it may do well to expose a list of fused_kernel_supported devices in this file that people can add their custom device to.